packages/sign-client/src/controllers/engine.ts

```diff
@@ -572,6 +591,402 @@ export class Engine extends IEngine {
+          namespaces: buildNamespacesFromAuth(
+            [...new Set(approvedMethods)],
+            [...new Set(approvedAccounts)],
+          ),
+        };
+
+        await this.client.core.relayer.subscribe(sessionTopic);
+        await this.client.session.set(sessionTopic, session);
+
+        session = this.client.session.get(sessionTopic);
+      }
+      resolve({
+        auths: cacaos,
+        session,
+      });
+    };
+
+    // set the ids for both requests
+    const id = payloadId();
+    const fallbackId = payloadId();
+
+    // subscribe to response events
+    this.events.once<"session_connect">(engineEvent("session_connect"), onSessionConnect);
+    this.events.once(engineEvent("session_request", id), onAuthenticate);
+
+    try {
+      // send both (main & fallback) requests
+      await Promise.all([
+        this.sendRequest({
+          topic: pairingTopic,
+          method: "wc_sessionAuthenticate",
+          params: request,
+          expiry: params.expiry,
+          throwOnFailedPublish: true,
+          clientRpcId: id,
+        }),
+        this.sendRequest({
+          topic: pairingTopic,
+          method: "wc_sessionPropose",
+          params: proposal,
+          expiry: ENGINE_RPC_OPTS.wc_sessionPropose.req.ttl,
+          throwOnFailedPublish: true,
+          clientRpcId: fallbackId,
+        }),
+      ]);
+    } catch (error) {
+      // cleanup listeners on failed publish
+      this.events.off(engineEvent("session_connect"), onSessionConnect);
+      this.events.off(engineEvent("session_request", id), onAuthenticate);
+      throw error;
+    }
+
+    await this.setProposal(fallbackId, { id: fallbackId, ...proposal });
+
+    await this.client.auth.requests.set(id, {
+      authPayload: request.authPayload,
+      requester: request.requester,
+      expiryTimestamp,
+      id,
+      pairingTopic,
+      verifyContext: {} as any,
+    });
+
+    return {
+      uri: connectionUri,
+      response: done,
+    } as EngineTypes.SessionAuthenticateResponsePromise;
+  };
+
+  public approveSessionAuthenticate: IEngine["approveSessionAuthenticate"] = async (
+    sessionAuthenticateResponseParams,
+  ) => {
+    this.isInitialized();
+
+    const { id, auths } = sessionAuthenticateResponseParams;
+
+    const pendingRequest = this.getPendingAuthRequest(id);
+
+    if (!pendingRequest) {
+      throw new Error(`Could not find pending auth request with id ${id}`);
+    }
+
+    const receiverPublicKey = pendingRequest.requester.publicKey;
+    const senderPublicKey = await this.client.core.crypto.generateKeyPair();
+    const responseTopic = hashKey(receiverPublicKey);
+
+    const encodeOpts = {
+      type: TYPE_1,
+      receiverPublicKey,
+      senderPublicKey,
+    };
+
+    const approvedMethods: string[] = [];
+    const approvedAccounts: string[] = [];
+    for (const cacao of auths) {
+      const isValid = await validateSignedCacao({ cacao, projectId: this.client.core.projectId });
```

Comment:
Since we're running the validation on `authenticate`, do we still need to validate `approveSessionAuthenticate` or is it safe to assume the proposal has valid cacaos ?
If we still need the validation, same suggestion to run this in parallel rather than sequentially.

Reply:
given that we need to respond if the validation fails, running in parallel with multiple invalid `cacaos` will result multiple error responces that can cause other potential issues. Lets keep it as is, and if we deem that performance is suffering will look into optimizing